### DISCUSSION TITLE
Trouble with particle times and variable extraction

### DISCUSSION THREAD

**Discussion by amargolskee** [Parcels-code/Parcels#1363] [Q&A]
Thanks in advance to anyone who helps with this. 

I'm having an issue with reading variables in at specific particle times. I want to evaluate variables for each particle at their initial seeding location at the their initial seeding time. I tried to follow one of the tutorials to do that.

First I read in x,y,z,t particle locations from a csv file. The times for particle seeding iterate from 1 to 67 in intervals of 6. I've checked that the fieldset time (set as fieldset.U.grid.time) produces particle times that are (0...2592000...51840360... and so on), as I expected.

However, when I run my code, I get this error:

```
INFO: Compiled ArraySampleParticleInitZeroSampleJNO3 ==> /tmp/parcels-347546/libb16df9b1eb30e586550f130776c8a1e9_0.so
WARNING: dt or runtime are zero, or endtime is equal to Particle.time. The kernels will be executed once, without incrementing time
Traceback (most recent call last):
  File "/scratch/gpfs/am9753/Parcels/Parcels_mpi/sose_parcels_d20.py", line 16, in <module>
    mitparc.run_mitgcm("jit")
  File "/scratch/gpfs/am9753/Parcels/Parcels_mpi/mitgcm_parcels_d20.py", line 81, in run_mitgcm
    pset.execute(
  File "/home/am9753/.conda/envs/py3_parcels_mpi3/lib/python3.11/site-packages/parcels/particleset/baseparticleset.py", line 450, in execute
    self.kernel.execute(self, endtime=next_time, dt=dt, recovery=recovery, output_file=output_file,
  File "/home/am9753/.conda/envs/py3_parcels_mpi3/lib/python3.11/site-packages/parcels/kernel/kernelsoa.py", line 227, in execute
    recovery_kernel(p, self.fieldset, p.time)
  File "/home/am9753/.conda/envs/py3_parcels_mpi3/lib/python3.11/site-packages/parcels/tools/statuscodes.py", line 221, in recovery_kernel_time_extrapolation
    raise OutOfTimeError(particle, fieldset)
parcels.tools.statuscodes.OutOfTimeError: 0
Particle P[311370](lon=3.333333, lat=-77.790306, depth=-2.100000, BLGBIONO3=0.000000, BLGMLD=0.000000, time=2592000.000000)
Time: 2013-02-04T00:00:00.000000000,	timestep dt: -0.000000
Field sampled outside time domain at time 2013-02-04T00:00:00.000000000. Try setting allow_time_extrapolation to True

```
2013-02-04 is definitely well within the runtime for the netcdf file that I'm reading from, so I'm very confused by this error. There are 511 5-day timesteps in the file. Here's a snippet of the netcdf structure. I would have thought that 2013-02-04 should just be 35 days into the model output. Is it necessary to set allow_time_extrapolation to True in this case? And if so, why?

It seems to run if I comment out the pset.execute() calls that have dt=0.. but it would be great to be able to get the initial variable reads to work at each particle's initial location.

```
netcdf bsose_Iter135_2013to2019_5day_BLGBIONO3 {
dimensions:
	time = 511 ;
	XC = 2160 ;
	YC = 588 ;
	Z = 52 ;
variables:
	int64 iter(time) ;
		iter:standard_name = "timestep" ;
		iter:long_name = "model timestep number" ;
	int64 time(time) ;
		time:standard_name = "time" ;
		time:long_name = "Time" ;
		time:axis = "T" ;
		time:units = "seconds since 2012-12-01" ;
		time:calendar = "proleptic_gregorian" ;

```

Here is the full file I'm using, in case that's helpful.

```
from datetime import timedelta as delta
from os import path
import sys
import numpy as np
import xarray as xr
import pandas as pd
import numpy.matlib
from parcels import (
    AdvectionRK4_3D,
    FieldSet,
    JITParticle,
    ParticleFile,
    ParticleSet,
    ScipyParticle,
    Variable,
    ErrorCode,
)

ptype = {"scipy": ScipyParticle, "jit": JITParticle}

class SampleParticleInitZero(JITParticle):         # Define a new particle class
    BLGBIONO3 = Variable('BLGBIONO3', initial=0)  # Variable 'JNO3' initialised by sampling the BLGBION
    BLGMLD = Variable('BLGMLD', initial=0)  # Variable 'MLD' initialised by sampling the BLGMLD


def SampleJNO3(particle, fieldset, time):
    particle.BLGBIONO3 = fieldset.BLGBIONO3[time, particle.depth, particle.lat, particle.lon]

def SampleMLD(particle, fieldset, time):
    particle.BLGMLD = fieldset.BLGMLD[time, 0, particle.lat, particle.lon]

def DeleteParticle(particle, fieldset, time):
    particle.delete()

def run_mitgcm(mode):
    """Function that shows how to load MITgcm data in a zonally periodic domain"""

    data_path = path.join(path.dirname(__file__), "/scratch/gpfs/am9753/SOSE/")
    filenames = {
        "U": data_path + "bsose_Iter135_2013to2019_5day_AllVel.nc",
        "V": data_path + "bsose_Iter135_2013to2019_5day_AllVel.nc",
        "W": data_path + "bsose_Iter135_2013to2019_5day_AllVel.nc",
        "BLGBIONO3": data_path + "bsose_Iter135_2013to2019_5day_BLGBIONO3.nc",
        "BLGMLD": data_path + "bsose_Iter135_2013to2019_5day_BLGMLD.nc"
    }
    variables = {"U": "u", "V": "v","W": "w", "BLGBIONO3": "BLGBIONO3","BLGMLD": "BLGMLD"}
    dimensions = {
        "U": {"lon": "XG", "lat": "YG", "time": "time","depth": "Z"},
        "V": {"lon": "XG", "lat": "YG", "time": "time","depth": "Z"},
        "W": {"lon": "XG", "lat": "YG", "time": "time","depth": "Z"},
        "BLGBIONO3": {"lon": "XC", "lat": "YC", "time": "time","depth": "Z"},
        "BLGMLD": {"lon": "XC", "lat": "YC", "time": "time"}
    }
    fieldset = FieldSet.from_mitgcm(filenames, variables,dimensions,
           allow_time_extrapolation=False)

#    data = pd.read_csv('init_loc.csv',header=None)
    data = pd.read_csv('init_loc_alld20xyt.csv',header=None)

    tmp = np.matrix(data)
    z = tmp[:,0]-1
    y = tmp[:,1]-1
    x = tmp[:,2]-1
    t = tmp[:,3]-1
    
    pset = ParticleSet(
         fieldset=fieldset,
         pclass=SampleParticleInitZero,
         lon = fieldset.U.grid.lon[x],
         lat = fieldset.U.grid.lat[y],
         depth = fieldset.U.grid.depth[z],
         time  = fieldset.U.grid.time[t],
    )

    pfile = ParticleFile(
        "SOSE_d20long_30" + ".zarr", pset, outputdt=delta(days=30)
    )
    sampleJNO3_kernel = pset.Kernel(SampleJNO3)
    sampleMLD_kernel = pset.Kernel(SampleMLD)
    kernels = AdvectionRK4_3D + sampleJNO3_kernel + sampleMLD_kernel
    pset.execute(
        sampleJNO3_kernel, dt=0,
        recovery={ErrorCode.ErrorOutOfBounds: DeleteParticle}
    )
    pset.execute(
        sampleMLD_kernel, dt=0,
        recovery={ErrorCode.ErrorOutOfBounds: DeleteParticle}
    )

    pset.execute(
        kernels, runtime=delta(days=2160), dt=delta(minutes=30), 
        output_file=pfile,
        recovery={ErrorCode.ErrorOutOfBounds: DeleteParticle} 
    )

    pfile.close()

```

**Comment by erikvansebille** [comment]
Thanks for reporting, @amargolskee. It could be that the calendar isn't parsed well when you create the fieldset, do you get any warnings about that? And what if you print `fieldset.U.grid.time_full`? What dates are then in the fieldset?

**Comment by amargolskee** [comment]
Thanks for getting back to me. The fieldset.U.grid.time_full and fieldset.U.grid.time dates appear identical.

Both are in units of seconds (0... 432000... up to 432000*511) and seemingly correspond to the amount of time since "2012-12-01" as in the netcdf file. Should these instead have been converted to datetime somewhere?

**Reply by erikvansebille** [reply]
Yes, they should indeed have been converted to a date time object; that apparently hasn't happened. It's probably beacuse Parcels doesn't know how to deal with the `time:calendar = "proleptic_gregorian" ;` in your netcdf file...

What is the value of `fieldset.U.grid.time_origin`?

**Comment by amargolskee** [comment]
I see. I didn't prepare these files myself: I am using B-SOSE output. The fielsdet.U.grid.time_origin is 2013-01-05T00:00:00.000000000

My python is a little shaky. Is there a quick fix I can use to convert the fieldset.U.grid.time* fields?

**Reply by erikvansebille** [reply]
Ah, the fact that the `fielsdet.U.grid.time_origin` is correct suggests that the date _is_ parsed well; so no problem with the calendar. In that case, I don't know what's going on with the time extrapolation error... Are you sure that you have data for that date?

**Reply by amargolskee** [reply]
Yeah... I'm also stumped. The first timestep I'm trying to read is the same as the first timestep in the netcdf file: 35 days after 2012-12-01. That's 2013-01-05, just like fieldset.U.grid.time_origin suggests. There's definitely data there and definitely data after it.

Even without turning allow_time_extrapolation to true, it runs if I comment out the dt=0 sampling kernels. I do want to be able to retrieve those initial conditions if possible. There's something fishy going on with the initial sampling for sure: it thinks there isn't data at the first timestep, but it doesn't complain for any timesteps after that.

I guess I could run it without the initial sampling and just do the interpolation outside of parcels during data analysis if there aren't any
other solutions to be had.

Thanks for walking through this with me.

**Reply by hadarberman1** [reply]
Hi, I am experiencing a similar issue, a time extrapolation error (Field sampled outside time domain at time 45885600.0. Try setting allow_time_extrapolation to True) when I issue a specific dt. If dt is commented the error does not appear. Has this issue been fixed? The time step is important in my run.

This is my code:

#%matplotlib inline
from parcels import FieldSet, ParticleSet, Variable, JITParticle, plotTrajectoriesFile, ScipyParticle, AdvectionRK4, ErrorCode
from Kernels import *
from ParticleClass import *
import numpy as np
import netCDF4 as nc
from datetime import timedelta as delta
import math
import matplotlib.pyplot as plt
import xarray as xr
import os
import sys
from glob import glob

#from datetime import timedelta
from operator import attrgetter
#fieldset = FieldSet.from_parcels("MovingEddies_data/moving_eddies")
#Coordinates of the test fieldset (on A-grid in deg)
filedir = './'
#zslice = sorted(glob(filedir + 'z_EMed300m_his.*.nc'))
zslice = filedir + 'z_EMed300m_his.00000.nc'
zslice0 = filedir + 'z_EMed300m_his.00000.nc'
grd = './EMed300m_grd.nc'

#load grid file
nchGD = nc.Dataset(grd, 'r')
lon_rho = nchGD.variables['lon_rho'][:]
lat_rho = nchGD.variables['lat_rho'][:]
mask_rho = nchGD.variables['mask_rho'][:]

filenames = {'U': {'lon': grd, 'lat': grd, 'depth': zslice0, 'data': zslice},
             'V': {'lon': grd, 'lat': grd, 'depth': zslice0, 'data': zslice}}

variables = {'U': 'u', 'V': 'v'}
#dimensions = {'U': {'lon': 'lon_psi', 'lat': 'lat_psi', 'depth': 'depth', 'time': 'time'},
#'V': {'lon': 'lon_psi', 'lat': 'lat_psi', 'depth': 'depth', 'time': 'time'}}
dimensions = {'U': {'time': 'time', 'depth': 'depth', 'lat': 'lat_psi', 'lon': 'lon_psi'},
              'V': {'time': 'time', 'depth': 'depth', 'lat': 'lat_psi', 'lon': 'lon_psi'}}




indices = {'time': range(0,3), 'depth' : [0], 'lat': range(0, 2173), 'lon': range(0, 1599)}


fieldset = FieldSet.from_c_grid_dataset(filenames, variables, dimensions, indices, allow_time_extrapolation=False)#, deferred_load=False)

#fieldset = FieldSet.from_mitgcm(filenames, variables, dimensions, indices, allow_time_extrapolation=False)#, deferred_load=False)
#U = 0*np.cos(0.55)*np.ones((lat.size, lon.size), dtype=np.float32) # no flow
#V = 0*np.sin(0.55)*np.ones((lat.size, lon.size), dtype=np.float32) # no flow
#PP = np.zeros((lat.size, lon.size), dtype=np.float32)
#mesh = 'flat'
#fieldset = FieldSet.from_data({'U': U, 'V': V, 'P': PP}, {'lon': lon, 'lat': lat}, mesh=mesh)
#fieldset = FieldSet.from_parcels("MovingEddies_data/moving_eddies")

#release particles in the boundary current
#lats1d = np.linspace(32.8479, 32.8487, num=200)
#lons1d = np.linspace(34.946, 33.951, num=200)

#[lons, lats] = np.meshgrid(lons1d, lats1d)
lats = 32.848+np.zeros((1,2),dtype=np.float32)
lons = 34.946+np.zeros((1,2),dtype=np.float32)
#[lons, lats] = np.meshgrid(lons1d, lats1d)
vector_size = lons.size
eps = np.random.rand(vector_size) * 1e-6  # Adjust the scaling factor as needed
lons = lons+eps
lats = lats+eps

nstart = 4
nend   = 8
nskip  = 4
ncycle = nend-nstart+nskip-1
print('ncycle: ', ncycle)

pset = ParticleSet.from_list(fieldset=fieldset, pclass=Hadar, lon=lons, lat=lats)
#pset = ParticleSet.from_list(fieldset=fieldset, pclass=ScipyParticle, lon=[3.3e5, 3.3e5], lat=[1e5, 2.8e5])
#k_WestVel = pset.Kernel(WestVel)        # casting the WestVel function to a kernel object
def DeleteParticle(particle, fieldset, time):
    particle.delete()

k_AdvectionRK4_HADAR = pset.Kernel(AdvectionRK4_HADAR_sphere)        # casting the WestVel function to a kernel object
k_AdvectionRK4 = pset.Kernel(AdvectionRK4)        # casting the WestVel function to a kernel object


output_file = pset.ParticleFile(name="Hadar_med_swim_Haifa4.zarr", outputdt=delta(minutes=1))

pset.execute(k_AdvectionRK4_HADAR,  # simply add kernels using the + operator
             runtime=delta(hours=8),
#dt=delta(seconds=0.1),
             output_file=output_file,recovery={ErrorCode.ErrorOutOfBounds: DeleteParticle})


With this Kernel:


def AdvectionRK4_HADAR_sphere(particle, fieldset, time):
    """Advection of particles using fourth-order Runge-Kutta integration.

    Function needs to be converted to Kernel object before execution"""
    import numpy as np

    # Parameters:
    # Time stepping parameters
    A = 0.05
    tau = particle.dt  # time step
    delta = np.sqrt(A * tau)  # turn clockwise
    # BCRW parameters
    sig02 = 0.0198  # angular diffusivity
    theta0 = 0.9035 * np.pi  # preferred direction
    B = 32.1968  # average time taken to reorient to the preferred direction

    # fixed parameters:
    muv = 0.1  # Omri's parameters
    sigmav = 0.03  # Omri's parameters
    if np.isnan(particle.theta):
        particle.theta = np.random.uniform(low=0.0, high=2 * np.pi, size=1)
    if np.isnan(particle.vswim):
        particle.vswim = np.random.normal(loc=muv, scale=sigmav, size=1)
    a = tau * (sig02 / (2 * delta ** 2) + (particle.theta - theta0) / (2 * B * delta))  # probability to turn clockwise
    b = tau * (sig02 / (2 * delta ** 2) - (particle.theta - theta0) / (
                2 * B * delta))  # probability to turn anticlockwise

    #     v=0
    # a[a < 0] = 0
    # a[a > 1] = 1
    # b[b < 0] = 0
    # b[b > 1] = 1

    c = 1 - a - b  # probability not to turn
    # c[c < 0] = 0

    Z = np.vstack((a, b, c))
    W_normalized = Z / np.tile(np.sum(Z, axis=0), (3, 1))
    W_cdf = np.cumsum(W_normalized, axis=0)
    xr = np.random.uniform(size=(1, Z.shape[1]))
    C = np.tile(xr, (3, 1)) < W_cdf
    idx = np.vstack((C[0, :], np.logical_xor(C[1:, :], C[:-1, :])))
    idxLocation = np.argmax(idx, axis=0)
    I = np.zeros_like(idxLocation)
    I[idxLocation == 0] = -1
    I[idxLocation == 1] = 1
    deltheta = delta * I
    particle.theta += deltheta

    (u1, v1) = fieldset.UV[particle]
    lon1, lat1 = (particle.lon + u1 * .5 * particle.dt, particle.lat + v1 * .5 * particle.dt)
    (u2, v2) = fieldset.UV[time + .5 * particle.dt, particle.depth, lat1, lon1, particle]
    lon2, lat2 = (particle.lon + u2 * .5 * particle.dt, particle.lat + v2 * .5 * particle.dt)
    (u3, v3) = fieldset.UV[time + .5 * particle.dt, particle.depth, lat2, lon2, particle]
    lon3, lat3 = (particle.lon + u3 * particle.dt, particle.lat + v3 * particle.dt)
    (u4, v4) = fieldset.UV[time + particle.dt, particle.depth, lat3, lon3, particle]
    particle.lon += (u1 + 2 * u2 + 2 * u3 + u4) / 6. * particle.dt + particle.vswim * tau * np.cos(
        particle.theta) / 1000. / 1.852 / 60.
    particle.lat += (v1 + 2 * v2 + 2 * v3 + v4) / 6. * particle.dt + particle.vswim * tau * np.sin(
        particle.theta) / 1000. / 1.852 / 60. / np.cos(particle.lat * np.pi / 180)


Thanks!

**Reply by erikvansebille** [reply]
I'm not sure that the difference between setting `dt=delta(seconds=0.1)` and using the default (which should be the same as using `dt=delta(seconds=1.0)`, can you check?) is not related to this issue. Could it be that this value of ` 45885600.0` is the end of your time series? I.e. is your input data  45885600.0 seconds long?

**Reply by hadarberman1** [reply]
Hi,

I'm not sure how to check the default dt value? Nevertheless, I'm sure that the default value differs from the specified one, since changing dt changes the runtime. 

I have found that for certain dt values there is no failure, i.e. dt=30s, dt=0.09s, and for other dt values the code fails at different time steps with the same time extrapolation error. It seems to me that this issue is related to the ratio of dt and 1 hr (which is the time interval between every two input files). 

Regarding your question, 45885600 is not the end of my input, and I am able to access the data with different dt values. 

While I can make slight adjustments to dt, I'm still concerned that it could encounter issues at some point due to the 'dt' to one-hour ratio. Does this make sense/ do you have any suggestions?

Thanks for your assistance,
Hadar.

**Reply by erikvansebille** [reply]
> I'm not sure how to check the default dt value? Nevertheless, I'm sure that the default value differs from the specified one, since changing dt changes the runtime.

Well, the default in parcels _is_ `dt=1.` seconds, see below
https://github.com/OceanParcels/parcels/blob/0265b5e73d475dd2f2bed64b84393872d56a591f/parcels/particleset.py#L803-L804

I'm not sure then what causes your error. Could you try updating to the new Parcels v3.0.0; in which the kernel loop has been rewritten. Perhaps that fixes the issue?

**Reply by hadarberman1** [reply]
Hi Erik,

Sure I can give it a try, how can I download the 3.0.0 version (using conda would be the best option for me)? Is it compatible with multi processing?

Thanks again,
Hadar.

**Reply by erikvansebille** [reply]
We released parcels v3 via condo-forge last week. However, some people are experiencing problems with installing (see https://github.com/conda-forge/parcels-feedstock/issues/100), so best way to install it is using (note that it can take a _long_ time for the repo data to be resolved)
```shell
conda create -n parcels3 -c conda-forge parcels=3
```
And then `parcels activate parcels3`
